sharktank/sharktank/utils/iree.py

```diff
@@ -753,3 +759,251 @@ def run_model_with_iree_run_module(
+
+    # empty tensors for export input
+    # there needs to be one corresponding to each arg
+    # NOTE: assuming args are not nested.
+    empty_args = tuple([torch.empty(arg.shape, dtype=arg.dtype) for arg in input_args])
+
+    # need to get this info from the test, currently only for static shapes
+    # one corresponding to each arg
+    dynamic_shapes = tuple([dict() for _ in input_args])
+
+    @fxb.export_program(
+        name=target_fn,
+        args=empty_args,
+        dynamic_shapes=(dynamic_shapes,),
+        strict=False,
+    )
+    def _(module, *fn_args):
+        return module.forward(*fn_args)
+
+    export_output = export(fxb, import_symbolic_shape_expressions=True)
+    export_output.save_mlir(mlir_path)
+
+    return expected, export_output
+
+
+def run_iree_module_from_vmfb(
+    vmfb_path: Path,
+    devices: list[iree.runtime.HalDevice] | None = None,
+    args=(),
+    *,
+    entrypoint="run_forward",
+    parameters_path=None,
+    driver="hip",
+    device_count=1,
+):
+    """
+    Load VMFB and run with IREE.
+
+    Args:
+        vmfb_path: Path to the VMFB file
+        args: Input arguments for the module
+        entrypoint: Name of the function to run
+        parameters_path: Optional path to parameters file
+        driver: IREE driver to use
+        device_count: Number of devices
+
+    Returns:
+        IREE module output
+    """
+    args = _as_tuple(args)
+
+    # Load & run with IREE
+    if devices is None:
+        devices = get_iree_devices(driver=driver, device_count=device_count)
+
+    def run_with_devices(devices):
+        iree_module, vm_context, _ = load_iree_module(
+            module_path=str(vmfb_path),
+            devices=devices,
+            parameters_path=parameters_path,
+        )
+        iree_args = prepare_iree_module_function_args(args=args, devices=devices)
+
+        iree_out = run_iree_module_function(
+            module=iree_module,
+            vm_context=vm_context,
+            args=iree_args,
+            device=devices[0],
+            function_name=entrypoint,
+        )
+        results = iree_to_torch(*iree_out)
+        return results
+
+    return with_iree_device_context(run_with_devices, devices)
+
+
+def compare_iree_torch_outputs(
+    iree_output,
+    torch_output,
+    *,
+    atol=1e-4,
+    rtol=0.0,
+):
+    """
+    Compare IREE output with torch eager reference and assert closeness.
+
+    Args:
+        iree_output: Output from IREE module as a tuple of np.ndarrays.
+        torch_output: Output from torch eager execution
+        atol/rtol: tolerances passed to torch.testing.assert_close
+    """
+    # Convert and compare
+    expected = torch_output
+    if isinstance(expected, torch.Tensor):
+        expected = (expected,)
+
+    actual = ()
+    for idx, o in enumerate(iree_output):
+        if isinstance(o, np.ndarray):
+            output = torch.tensor(o).type_as(expected[idx])
+        else:
+            output = o
```

Comment:
I would expect that this would use the existing conversion functions for IREE device array to torch tensor before doing the comparison.